polaris-service/src/main/java/org/apache/polaris/service/ratelimiter/RateLimiterFactory.java

```diff
@@ -0,0 +1,34 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.polaris.service.ratelimiter;
+
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import io.dropwizard.jackson.Discoverable;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Interface for constructing a rate limiter given the rate limiting key and clock. Notably, rate
+ * limiter construction may be asynchronous. This allows fetching information related to the key.
+ * For example, implementors may wish to fetch per-account rate limit values, which require lookup
+ * in an external database.
+ */
+@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "type")
+public interface RateLimiterFactory extends Discoverable {
+  CompletableFuture<RateLimiter> createRateLimiter(String key, Clock clock);
```

Comment:
Instead of a `CompleteableFuture` maybe we can just return a `Future` / `Promise` -- if the caller needs to check when it's complete, it could also be a `CompletionStage`

Reply:
Updated